### PR TITLE
Allow access the table while altering distribution policy.

### PR DESCRIPTION
**Pull request by kainwen**
With postgres9.4 merged, we have catalog mvcc, so it is
OK to only hold ExclusiveLock on the table while altering
its distribution policy.

This commit also removes the reshuffle tests under regress
because they can all be covered in isolation2's test.

### CODE REVIEW

**Comment by d** [comment]
The test looks a lot more proper in this patch than in PR #5896 , because `SELECT a FROM replicated_foo` doesn't reveal the actual data on all segments.

While not the fault of this PR, why is `gp_distribution_policy` not a replicated table? I see that you need to hack the local replica of that table on each segment.

In fact, this hacking around seems to be indicative of a UX gap: why isn't `numsegments` controlled by the `CREATE TABLE` DDL? How can the change in #5896 properly enable table contraction (moving from 0,1,2,3,4 to 0,1,2)

**Comment by kainwen** [comment]
@d 
Thanks for your commenting.
* `create table` DDL always set `numsegments` to current cluster size
* In practice, `numsegments` can (should) only be updated by `gpexpand`, for the ease of the test, we hack this way.

**Comment by pivotal-ning-yu** [comment]
LGTM, except some minor comments:
1. is the `setdistributedby` dir really necessary? it's so long and the test file already has a descriptive name;
2. better to also add the case for concurrent ALTER TABLE (or maybe we already have one?).

File: src/test/isolation2/sql/reshuffle.sql [hunk 1]
```diff
@@ -0,0 +1,738 @@
...
+1: begin;
+1: alter table thash_reshuffle set with (reshuffle);
+
+-- since alter reshuffle transaction has not committed
+-- the other transaction can access the table, but
+-- can only see tuples from seg0, seg1.
+select * from thash_reshuffle order by a, b;
+select 2 in (select gp_segment_id from thash_reshuffle) as check;
+
+29: begin;
```
Comment (kainwen) [inline src/test/isolation2/sql/reshuffle.sql]: @pivotal-ning-yu I add some cases to test concurrently modifying tables. 
Just here, I think it is enough.

File: src/test/isolation2/sql/reshuffle.sql [hunk 2]
```diff
@@ -0,0 +1,738 @@
...
+31q:
+
+select * from thash_reshuffle order by a, b;
+select 2 in (select gp_segment_id from thash_reshuffle) as check;
+
+1: begin;
+1: alter table thash_reshuffle set with (reshuffle);
+
+32: begin;
+32&: alter table thash_reshuffle set with (reshuffle);
```
Comment (kainwen) [inline src/test/isolation2/sql/reshuffle.sql]: @pivotal-ning-yu 
I add a case of concurrently alter reshuffle here. 
I think it is enough only test here.

File: src/test/isolation2/sql/reshuffle.sql [hunk 3]
```diff
@@ -0,0 +1,738 @@
...
+3: alter table trep_reshuffle set with (reshuffle);
+
+select * from trep_reshuffle order by a, b;
+select select_on_segment('select * from trep_reshuffle', 2);
+
+3: commit;
+3q:
+
+select * from trep_reshuffle order by a, b;
+-- select select_on_segment('select * from trep_reshuffle', 2);
```
Comment (kainwen) [inline src/test/isolation2/sql/reshuffle.sql]: Let me uncomment this to truely test replicated table's reshuffle after the PR is merged: https://github.com/greenplum-db/gpdb/pull/6069

**Comment by kainwen** [comment]
@hlinnaka 
Thanks.

1. Not all other comments take AccessExclusiveLock. Some subcommands even allow modifying the table.
2. The reshuffle command here, is almost the same as an update statement on hash columns, the only difference is that reshuffle command also update the catalog `gp_distribution_policy` at the last step. So it is safe for other transactions access the tables?
3. I don't quite understand that with catalog mvcc enabled, why we take the highest level lock for many alter table commands? The comments say some command will rewrite heap and may be visible. I myself do not know what case will cause this? Would you please help solve this doubts?

Thank you again in advance!

**Comment by kainwen** [comment]
@d 
We have not thought of shrinking the cluster.

**Comment by hlinnaka** [comment]
On 25 October 2018 07:14:57 WEST, Zhenghua Lyu <notifications@github.com> wrote:
>@hlinnaka 
>Thanks.
>
>1. Not all other comments take AccessExclusiveLock. Some subcommands
>even allow modifying the table.
>2. The reshuffle command here, is almost the same as an update
>statement on hash columns, the only difference is that reshuffle
>command also update the catalog `gp_distribution_policy` at the last
>step. So it is safe for other transactions access the tables?

Ah, ok, I thought it's like the table-rewriting variants of ALTER TABLE.

>3. I don't quite understand that with catalog mvcc enabled, why we take
>the highest level lock for many alter table commands? The comments say
>some command will rewrite heap and may be visible. I myself do not know
>what case will cause this? Would you please help solve this doubts?

We use MVCC snapshots for catalog scans now, but it's still different from the snapshot we use for visibility checks on data. We take a new MVCC snapshot for each catalog access (although I think there's some caching involved too), and the catalog snapshot is a local snapshot that ignores distributed visibility.

The lock level for some of those commands is actually relaxed in later PostgreSQL versions, so I think 9.4 was indeed overly conservative.

I'm not sure if this patch is correct or not, but it seems suspicious.


- Heikki

**Comment by kainwen** [comment]
The lock level cannot be reduced there are many risks. Close this PR.

Will re-open and refine it in future if it is possible.

**Comment by kainwen** [comment]
@hlinnaka 

The lock level cannot be reduced currently. 

There are several critial stages during a query's lifetime:
1. make the plan
2. PortalStart (at this step, it will get snapshot for data scan)
3. AssignGangs
4. Exec & finish...

If a reshuffle transaction (named `Txn_resfhulle`) commits just between step 2 and step3, which means:
1. we make plan as numsegments = 2 (with `Txn_resfhulle` uncommitted)
2. we scan data using a snapshot (with `Txn_resfhulle` committed)

Then we will only create a gang with size = 2, and miss some data.

**Pull request closed without merging**